### PR TITLE
Add support for output layout NHWC

### PR DESCRIPTION
**Pull request by kgrundy20**
Support for output NHWC layout that covers milestone 4 workloads.

### CODE REVIEW

File: tools/cross_runner/src/conv.h [hunk 1]
```diff
@@ -9,16 +9,31 @@ class Convolution : public DirectMlBaseNode
...
+        bool use_bias, bool allow_fp16_computations,
+        IDMLDevice* dml_device, ID3D12Device* d3d12_device, bool disable_mc = false)
         : DirectMlBaseNode(dml_device, d3d12_device)
     {
 
         const dml::TensorDimensions input_dims{ input_shape.n, input_shape.c, input_shape.h, input_shape.w };
         const dml::TensorDimensions filter_dims{ filter_shape.n, filter_shape.c, filter_shape.h, filter_shape.w };
-        const dml::TensorDimensions output_dims{ output_shape.n, output_shape.c, output_shape.h, output_shape.w };
+        dml::TensorDimensions output_dims;
+        if (output_shape.nhwc)
```
Comment (smarcink) [inline tools/cross_runner/src/conv.h:22]: This doesnt seem right.
Tensor is always in order: NCHW
what determines the layout should be strides calculated object: TensorPolicy output_tensor_policy

File: tools/cross_runner/src/layers_utils.h [hunk 2]
```diff
@@ -23,11 +23,12 @@ struct TensorShape
     std::uint32_t d = 0; // for 5d tensors
     std::uint32_t h = 0;
     std::uint32_t w = 0;
+    bool nhwc = 0;
```
Comment (smarcink) [inline tools/cross_runner/src/layers_utils.h:26]: Tensor shape should be layout agnostic
It's just shape (dimensions)
It shouldnt know about layout.

File: tools/cross_runner/kernels/conv_nchw_fp16.cpp [hunk 3]
```diff
@@ -290,13 +303,22 @@ extern "C" _GENX_MAIN_ void convolution_nchw_nondpas(
...
 	const uint output_batch_offset = batch_id * BLOCK_BATCH * output_height * output_width * output_channels;
-    const uint output_oc_chunk_offset = oc_chunk_id * BLOCK_OC * output_height * output_width;
-    const uint output_w_chunk_offset = w_chunk_id * BLOCK_W;
-    const uint output_h_chunk_offset = h_chunk_id * BLOCK_H * output_width;
-    uint32_t output_offset = (output_batch_offset + output_oc_chunk_offset + output_h_chunk_offset + output_w_chunk_offset) * sizeof(DT_OUT);
-	
-	store_output_wc8_as_nchw(surface_output, output_row_0.row(0), output_width, output_height, output_offset, w_chunk_id);
-	//store_output_wc8_as_nchw(surface_output, output_row_0.row(1), output_width, output_height, output_offset + output_height * output_width * output_channels * sizeof(DT_OUT), w_chunk_id);
-}
+#if OUTPUT_LAYOUT_IS_NHWC
```
Comment (smarcink) [inline tools/cross_runner/kernels/conv_nchw_fp16.cpp:308]: What is performance impact if this is runtime condition vs compile time condition?
Could you collect some perf data?

File: tools/cross_runner/kernels/conv_nchw_fp16.cpp [hunk 4]
```diff
@@ -205,6 +205,19 @@ _GENX_ inline void store_output_wc8_as_nchw(SurfaceIndex surface [[type("buffer_
 	}
 }
 
+_GENX_ inline void store_output_wc8_as_nhwc(SurfaceIndex surface [[type("buffer_t")]], vector_ref<DT_OUT, ACCU_REG_SIZE> grf_chunk, uint32_t output_width, uint32_t output_channels, uint32_t byte_offset)
+{
+	const uint32_t BLOCK_SC = MAX_ELEMENT_SIZE / OUTPUT_ELEMENT_SIZE;
+
+	for (int i = 0; i < BLOCK_W; i++)
```
Comment (smarcink) [inline tools/cross_runner/kernels/conv_nchw_fp16.cpp:212]: `#pragma unroll`   can help with performance if BLOCK_W is compile time constant

**Review by smarcink — CHANGES_REQUESTED** [review]

File: tools/cross_runner/kernels/conv_nchw_fp16.cpp (hunk 3 above)
Comment (kgrundy20) [reply tools/cross_runner/kernels/conv_nchw_fp16.cpp:310]: The runtime condition is faster, so I'll make that change

File: tools/cross_runner/kernels/conv_nchw_fp16.cpp (hunk 4 above)
Comment (kgrundy20) [reply tools/cross_runner/kernels/conv_nchw_fp16.cpp:214]: Added

File: tools/cross_runner/kernels/conv_nchw_fp16.cpp [hunk 5]
```diff
@@ -242,6 +256,7 @@ extern "C" _GENX_MAIN_ void convolution_nchw_nondpas(
 	const uint32_t activation_type = constants[9];
 	const uint32_t activation_alpha = constants[10];
 	const uint32_t activation_beta = constants[11];
+	const uint32_t output_layout = constants[12];
```
Comment (smarcink) [inline tools/cross_runner/kernels/conv_nchw_fp16.cpp:259]: Please rename it to be a bit more meaningful, for example: `output_layout_is_nhwc`

**Review by smarcink — APPROVED** [review]
LGTM, added minor comment

**Pull request merged**